// awesome-typescript-loader/src/cache.js
var fs = require('fs');
var crypto = require('crypto');
var os = require('os');
var path = require('path');
var zlib = require('zlib');
function findCompiledModule(fileName) {
    var baseFileName = fileName.replace(/(\.ts|\.tsx)$/, '');
    var compiledFileName = baseFileName + ".js";
    if (fs.existsSync(compiledFileName)) {
        var mapFileName = baseFileName + ".js.map";
        var isMapExists = fs.existsSync(mapFileName);
        var result = {
            fileName: compiledFileName,
            text: fs.readFileSync(compiledFileName).toString(),
            mapName: isMapExists
                ? mapFileName
                : null,
            map: isMapExists
                ? fs.readFileSync(mapFileName).toString()
                : null
        };
        return result;
    }
    else {
        return null;
    }
}
exports.findCompiledModule = findCompiledModule;
/**
 * Read the contents from the compressed file.
 *
 * @async
 */
function read(filename, callback) {
    return fs.readFile(filename, function (err, data) {
        if (err) {
            return callback(err);
        }
        return zlib.gunzip(data, function (err, content) {
            var result = {};
            if (err) {
                return callback(err);
            }
            try {
                result = JSON.parse(content);
            }
            catch (e) {
                return callback(e);
            }
            return callback(null, result);
        });
    });
}
;
/**
 * Write contents into a compressed file.
 *
 * @async
 * @params {String} filename
 * @params {String} result
 * @params {Function} callback
 */
function write(filename, result, callback) {
    var content = JSON.stringify(result);
    return zlib.gzip(content, function (err, data) {
        if (err) {
            return callback(err);
        }
        return fs.writeFile(filename, data, callback);
    });
}
;
/**
 * Build the filename for the cached file
 *
 * @params {String} source  File source code
 * @params {Object} options Options used
 *
 * @return {String}
 */
function filename(source, identifier, options) {
    var hash = crypto.createHash('sha512');
    var contents = JSON.stringify({
        identifier: identifier,
        options: options,
        source: source,
    });
    hash.end(contents);
    return hash.read().toString('hex') + '.json.gzip';
}
;
/**
 * Retrieve file from cache, or create a new one for future reads
 *
 * @async
 * @example
 */
function cache(params, callback) {
    // Spread params into named variables
    // Forgive user whenever possible
    var source = params.source;
    var options = params.options || {};
    var transform = params.transform;
    var identifier = params.identifier;
    var directory = (typeof params.directory === 'string') ?
        params.directory :
        os.tmpdir();
    var file = path.join(directory, filename(source, identifier, options));
    return read(file, function (err, content) {
        var result = {};
        // No errors mean that the file was previously cached
        // we just need to return it
        if (!err) {
            return callback(null, content);
        }
        // Otherwise just transform the file
        // return it to the user asap and write it in cache
        try {
            result = transform(source, options);
        }
        catch (error) {
            return callback(error);
        }
        return write(file, result, function (err) {
            return callback(err, result);
        });
    });
}
exports.cache = cache;
;
