// angular2-datatable/lib/DataTable.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var PageData = (function () {
    function PageData() {
    }
    return PageData;
}());
exports.PageData = PageData;
var DataTable = (function () {
    function DataTable() {
        this.inputData = new PageData();
        this.sortBy = "";
        this.sortOrder = "asc";
        this.rowsOnPage = 1000;
        this.activePage = 1;
        this.mustRecalculateData = false;
        this.pageData = new PageData();
        this.onDataChange = new core_1.EventEmitter();
        this.onSortChange = new core_1.EventEmitter();
        this.onPageChange = new core_1.EventEmitter();
        this.fpage = new core_1.EventEmitter();
    }
    DataTable.prototype.getSort = function () {
        return { sortBy: this.sortBy, sortOrder: this.sortOrder };
    };
    DataTable.prototype.setSort = function (sortBy, sortOrder) {
        if (this.sortBy !== sortBy || this.sortOrder !== sortOrder) {
            this.sortBy = sortBy;
            this.sortOrder = sortOrder;
            this.mustRecalculateData = true;
            this.onSortChange.emit({ sortBy: sortBy, sortOrder: sortOrder });
        }
    };
    DataTable.prototype.getPage = function () {
        return { activePage: this.inputData.page, rowsOnPage: this.inputData.size, dataLength: this.inputData.total };
    };
    DataTable.prototype.setPage = function (activePage, rowsOnPage) {
        if (this.inputData.size !== rowsOnPage || this.inputData.page !== activePage) {
            this.inputData.page = this.inputData.page !== activePage ? activePage : this.calculateNewActivePage(this.inputData.size, rowsOnPage);
            this.inputData.size = rowsOnPage;
            this.mustRecalculateData = true;
            this.onPageChange.emit({ activePage: this.inputData.page, rowsOnPage: this.inputData.size, dataLength: this.inputData.total });
            this.fpage.emit(this.pageData);
        }
    };
    DataTable.prototype.calculateNewActivePage = function (previousRowsOnPage, currentRowsOnPage) {
        var firstRowOnPage = (this.inputData.page - 1) * previousRowsOnPage + 1;
        var newActivePage = Math.ceil(firstRowOnPage / currentRowsOnPage);
        return newActivePage;
    };
    DataTable.prototype.ngOnChanges = function (changes) {
        if (changes["inputData"]) {
            this.inputData = this.inputData || new PageData();
            this.onPageChange.emit({
                activePage: this.inputData.page,
                rowsOnPage: this.inputData.size,
                dataLength: this.inputData.total
            });
            this.mustRecalculateData = true;
        }
    };
    DataTable.prototype.ngDoCheck = function () {
        if (this.mustRecalculateData) {
            this.fillData();
            this.mustRecalculateData = false;
        }
    };
    DataTable.prototype.fillData = function () {
        this.inputData.page = this.inputData.page;
        this.inputData.size = this.inputData.size;
        var offset = (this.inputData.page - 1) * this.inputData.size;
        this.pageData.data = this.inputData.data;
        this.pageData.iDisplayStart = offset;
        this.pageData.iDisplayLength = this.inputData.size;
        this.pageData.page = this.inputData.page;
        this.pageData.size = this.inputData.size;
        this.data = this.pageData;
    };
    __decorate([
        core_1.Input("mfData"), 
        __metadata('design:type', PageData)
    ], DataTable.prototype, "inputData", void 0);
    __decorate([
        core_1.Input("mfRowsOnPage"), 
        __metadata('design:type', Object)
    ], DataTable.prototype, "rowsOnPage", void 0);
    __decorate([
        core_1.Input("mfActivePage"), 
        __metadata('design:type', Object)
    ], DataTable.prototype, "activePage", void 0);
    DataTable = __decorate([
        core_1.Directive({
            selector: 'table[mfData]',
            outputs: ['fpage'],
            exportAs: 'mfDataTable'
        }), 
        __metadata('design:paramtypes', [])
    ], DataTable);
    return DataTable;
}());
exports.DataTable = DataTable;
