// es7-reflect-metadata/src/shim/weakmap.js
var create_unique_key_1 = require("../helper/create-unique-key");
var get_or_create_weakmap_1 = require("../helper/get-or-create-weakmap");
exports.rootKey = create_unique_key_1.createUniqueKey();
var WeakMap = (function () {
    function WeakMap() {
        this._key = create_unique_key_1.createUniqueKey();
    }
    Object.defineProperty(WeakMap.prototype, "length", {
        get: function () {
            return 0;
        },
        enumerable: true,
        configurable: true
    });
    WeakMap.prototype.has = function (target) {
        var table = get_or_create_weakmap_1.getOrCreateWeakMapTable(exports.rootKey, target, /*create*/ false);
        if (table) {
            return this._key in table;
        }
        return false;
    };
    WeakMap.prototype.get = function (target) {
        var table = get_or_create_weakmap_1.getOrCreateWeakMapTable(exports.rootKey, target, /*create*/ false);
        if (table) {
            return table[this._key];
        }
        return undefined;
    };
    WeakMap.prototype.set = function (target, value) {
        var table = get_or_create_weakmap_1.getOrCreateWeakMapTable(exports.rootKey, target, /*create*/ true);
        table[this._key] = value;
        return this;
    };
    WeakMap.prototype.delete = function (target) {
        var table = get_or_create_weakmap_1.getOrCreateWeakMapTable(exports.rootKey, target, /*create*/ false);
        if (table && this._key in table) {
            return delete table[this._key];
        }
        return false;
    };
    WeakMap.prototype.clear = function () {
        // NOTE: not a real clear, just makes the previous data unreachable
        this._key = create_unique_key_1.createUniqueKey();
    };
    return WeakMap;
})();
exports.WeakMap = WeakMap;
function createWeakMapPolyfill() {
    return WeakMap;
}
exports.createWeakMapPolyfill = createWeakMapPolyfill;
