// ng2-validate/bundles/ng2-validate.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
System.register("core/models/base.validation", [], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var BaseValidation;
    return {
        setters:[],
        execute: function() {
            BaseValidation = (function () {
                function BaseValidation(config) {
                    if (!config) {
                        this.config = this.getDefaultConfig();
                    }
                    else {
                        this.config = config;
                        this.resolveConfig(this.config, this.getDefaultConfig());
                    }
                }
                BaseValidation.prototype.resolveConfig = function (preferredConfig, defaultConfig) {
                    for (var key in preferredConfig) {
                        preferredConfig[key] = this.resolveValue(preferredConfig[key], defaultConfig[key]);
                    }
                };
                BaseValidation.prototype.resolveValue = function (preferredValue, defaultValue) {
                    if (preferredValue === undefined) {
                        return defaultValue;
                    }
                    return preferredValue;
                };
                return BaseValidation;
            }());
            exports_1("BaseValidation", BaseValidation);
        }
    }
});
System.register("core/models/validation-message.model", [], function(exports_2, context_2) {
    "use strict";
    var __moduleName = context_2 && context_2.id;
    var ValidationMessage;
    return {
        setters:[],
        execute: function() {
            ValidationMessage = (function () {
                function ValidationMessage(message, context) {
                    this.text = message;
                    this.context = context;
                }
                return ValidationMessage;
            }());
            exports_2("ValidationMessage", ValidationMessage);
        }
    }
});
System.register("core/models/field-validator-result.interface", [], function(exports_3, context_3) {
    "use strict";
    var __moduleName = context_3 && context_3.id;
    return {
        setters:[],
        execute: function() {
        }
    }
});
System.register("core/models/field-validation.interface", [], function(exports_4, context_4) {
    "use strict";
    var __moduleName = context_4 && context_4.id;
    return {
        setters:[],
        execute: function() {
        }
    }
});
System.register("core/models/validation-messages.model", ["core/models/validation-message.model"], function(exports_5, context_5) {
    "use strict";
    var __moduleName = context_5 && context_5.id;
    var validation_message_model_1;
    var ValidationMessages;
    return {
        setters:[
            function (validation_message_model_1_1) {
                validation_message_model_1 = validation_message_model_1_1;
            }],
        execute: function() {
            ValidationMessages = (function () {
                function ValidationMessages() {
                    this._messages = {};
                }
                ValidationMessages.prototype.addMessage = function (key, message, context) {
                    this._messages[key] = new validation_message_model_1.ValidationMessage(message, context);
                };
                ValidationMessages.prototype.resolve = function () {
                    if (Object.keys(this._messages).length === 0) {
                        return null;
                    }
                    return Object.assign({}, this._messages);
                };
                return ValidationMessages;
            }());
            exports_5("ValidationMessages", ValidationMessages);
        }
    }
});
System.register("core/components/validation-message/validation-message.component", ['@angular/core'], function(exports_6, context_6) {
    "use strict";
    var __moduleName = context_6 && context_6.id;
    var core_1;
    var ValidationMessageComponent;
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            }],
        execute: function() {
            ValidationMessageComponent = (function () {
                function ValidationMessageComponent() {
                }
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', String)
                ], ValidationMessageComponent.prototype, "message", void 0);
                ValidationMessageComponent = __decorate([
                    core_1.Component({
                        selector: 'validation-message',
                        template: "\n    <section *ngIf=\"message\">\n        {{message}}\n    </section>\n    "
                    }), 
                    __metadata('design:paramtypes', [])
                ], ValidationMessageComponent);
                return ValidationMessageComponent;
            }());
            exports_6("ValidationMessageComponent", ValidationMessageComponent);
        }
    }
});
System.register("core/components/validation-messages/validation-messages.component", ['@angular/core', '@angular/common', "core/components/validation-message/validation-message.component"], function(exports_7, context_7) {
    "use strict";
    var __moduleName = context_7 && context_7.id;
    var core_2, common_1, validation_message_component_1;
    var ValidationMessagesComponent;
    return {
        setters:[
            function (core_2_1) {
                core_2 = core_2_1;
            },
            function (common_1_1) {
                common_1 = common_1_1;
            },
            function (validation_message_component_1_1) {
                validation_message_component_1 = validation_message_component_1_1;
            }],
        execute: function() {
            ValidationMessagesComponent = (function () {
                function ValidationMessagesComponent() {
                }
                ValidationMessagesComponent.prototype.ngOnInit = function () {
                    this.messageProperty = this.messageProperty || 'text';
                };
                Object.defineProperty(ValidationMessagesComponent.prototype, "messages", {
                    get: function () {
                        return this.filterMessages();
                    },
                    enumerable: true,
                    configurable: true
                });
                ValidationMessagesComponent.prototype.filterMessages = function () {
                    var result = null;
                    if (this.control && this.control.errors && Object.keys(this.control.errors).length > 0) {
                        result = [];
                        var exemptions = this.exemptKeys ? this.exemptKeys : [];
                        var _loop_1 = function(key) {
                            if (!exemptions.find(function (e) { return e.toLowerCase() === key.toLowerCase(); })) {
                                result.push(this_1.control.errors[key]);
                            }
                        };
                        var this_1 = this;
                        for (var key in this.control.errors) {
                            _loop_1(key);
                        }
                        if (result.length === 0) {
                            result = null;
                        }
                    }
                    return result;
                };
                __decorate([
                    core_2.Input(), 
                    __metadata('design:type', common_1.AbstractControl)
                ], ValidationMessagesComponent.prototype, "control", void 0);
                __decorate([
                    core_2.Input(), 
                    __metadata('design:type', Array)
                ], ValidationMessagesComponent.prototype, "exemptKeys", void 0);
                __decorate([
                    core_2.Input(), 
                    __metadata('design:type', String)
                ], ValidationMessagesComponent.prototype, "messageProperty", void 0);
                ValidationMessagesComponent = __decorate([
                    core_2.Component({
                        selector: 'validation-messages',
                        template: "\n    <div *ngIf=\"messages\" *ngFor=\"#message of messages\">\n        <validation-message *ngIf=\"message\" [message]=\"message[messageProperty]\"></validation-message>\n    </div>",
                        directives: [validation_message_component_1.ValidationMessageComponent]
                    }), 
                    __metadata('design:paramtypes', [])
                ], ValidationMessagesComponent);
                return ValidationMessagesComponent;
            }());
            exports_7("ValidationMessagesComponent", ValidationMessagesComponent);
        }
    }
});
System.register("core", ["core/models/base.validation", "core/models/validation-message.model", "core/models/validation-messages.model", "core/components/validation-message/validation-message.component", "core/components/validation-messages/validation-messages.component"], function(exports_8, context_8) {
    "use strict";
    var __moduleName = context_8 && context_8.id;
    return {
        setters:[
            function (base_validation_1_1) {
                exports_8({
                    "BaseValidation": base_validation_1_1["BaseValidation"]
                });
            },
            function (validation_message_model_2_1) {
                exports_8({
                    "ValidationMessage": validation_message_model_2_1["ValidationMessage"]
                });
            },
            function (validation_messages_model_1_1) {
                exports_8({
                    "ValidationMessages": validation_messages_model_1_1["ValidationMessages"]
                });
            },
            function (validation_message_component_2_1) {
                exports_8({
                    "ValidationMessageComponent": validation_message_component_2_1["ValidationMessageComponent"]
                });
            },
            function (validation_messages_component_1_1) {
                exports_8({
                    "ValidationMessagesComponent": validation_messages_component_1_1["ValidationMessagesComponent"]
                });
            }],
        execute: function() {
        }
    }
});
System.register("logic/logics/hasValue.logic", ['@angular/core/src/facade/lang'], function(exports_9, context_9) {
    "use strict";
    var __moduleName = context_9 && context_9.id;
    var lang_1;
    var HasValue;
    return {
        setters:[
            function (lang_1_1) {
                lang_1 = lang_1_1;
            }],
        execute: function() {
            HasValue = (function () {
                function HasValue() {
                }
                HasValue.check = function (control) {
                    if (lang_1.isBlank(control.value) || (lang_1.isString(control.value) && control.value.trim() === '')) {
                        return false;
                    }
                    return true;
                };
                return HasValue;
            }());
            exports_9("HasValue", HasValue);
        }
    }
});
System.register("logic/logics/numeric/isNumeric.logic", [], function(exports_10, context_10) {
    "use strict";
    var __moduleName = context_10 && context_10.id;
    var IsNumeric;
    return {
        setters:[],
        execute: function() {
            IsNumeric = (function () {
                function IsNumeric() {
                }
                IsNumeric.check = function (control) {
                    if (control.value === undefined || control.value === null || control.value instanceof Object || isNaN(+control.value) === true) {
                        return false;
                    }
                    return true;
                };
                return IsNumeric;
            }());
            exports_10("IsNumeric", IsNumeric);
        }
    }
});
System.register("logic/logics/numeric/min.logic", [], function(exports_11, context_11) {
    "use strict";
    var __moduleName = context_11 && context_11.id;
    var Min;
    return {
        setters:[],
        execute: function() {
            Min = (function () {
                function Min() {
                }
                Min.check = function (control, min) {
                    if (isNaN(+control.value) === true) {
                        throw new Error("\"" + control.value + "\" is not a number\"");
                    }
                    if (+control.value < min) {
                        return false;
                    }
                    return true;
                };
                return Min;
            }());
            exports_11("Min", Min);
        }
    }
});
System.register("logic/logics/numeric/max.logic", [], function(exports_12, context_12) {
    "use strict";
    var __moduleName = context_12 && context_12.id;
    var Max;
    return {
        setters:[],
        execute: function() {
            Max = (function () {
                function Max() {
                }
                Max.check = function (control, max) {
                    if (isNaN(+control.value) === true) {
                        throw new Error("\"" + control.value + "\" is not a number\"");
                    }
                    if (+control.value > max) {
                        return false;
                    }
                    return true;
                };
                return Max;
            }());
            exports_12("Max", Max);
        }
    }
});
System.register("logic/logics/numeric/range.logic", [], function(exports_13, context_13) {
    "use strict";
    var __moduleName = context_13 && context_13.id;
    var Range;
    return {
        setters:[],
        execute: function() {
            Range = (function () {
                function Range() {
                }
                Range.check = function (control, minValue, maxValue) {
                    if (isNaN(+control.value) === true) {
                        throw new Error("\"" + control.value + "\" is not a number\"");
                    }
                    if (+control.value < minValue || +control.value > maxValue) {
                        return false;
                    }
                    return true;
                };
                return Range;
            }());
            exports_13("Range", Range);
        }
    }
});
System.register("logic/logics/length/minLength.logic", [], function(exports_14, context_14) {
    "use strict";
    var __moduleName = context_14 && context_14.id;
    var MinLength;
    return {
        setters:[],
        execute: function() {
            MinLength = (function () {
                function MinLength() {
                }
                MinLength.check = function (control, minLength) {
                    if (control.value.length < minLength) {
                        return false;
                    }
                    return true;
                };
                return MinLength;
            }());
            exports_14("MinLength", MinLength);
        }
    }
});
System.register("logic/logics/length/maxLength.logic", [], function(exports_15, context_15) {
    "use strict";
    var __moduleName = context_15 && context_15.id;
    var MaxLength;
    return {
        setters:[],
        execute: function() {
            MaxLength = (function () {
                function MaxLength() {
                }
                MaxLength.check = function (control, maxLength) {
                    if (control.value.length > maxLength) {
                        return false;
                    }
                    return true;
                };
                return MaxLength;
            }());
            exports_15("MaxLength", MaxLength);
        }
    }
});
System.register("logic/logics/length/length.logic", [], function(exports_16, context_16) {
    "use strict";
    var __moduleName = context_16 && context_16.id;
    var Length;
    return {
        setters:[],
        execute: function() {
            Length = (function () {
                function Length() {
                }
                Length.check = function (control, minLength, maxLength) {
                    if (control.value.length < minLength || control.value.length > maxLength) {
                        return false;
                    }
                    return true;
                };
                return Length;
            }());
            exports_16("Length", Length);
        }
    }
});
System.register("logic/logics/length/isLength.logic", [], function(exports_17, context_17) {
    "use strict";
    var __moduleName = context_17 && context_17.id;
    var IsLength;
    return {
        setters:[],
        execute: function() {
            IsLength = (function () {
                function IsLength() {
                }
                IsLength.check = function (control, length) {
                    if (control.value.length !== length) {
                        return false;
                    }
                    return true;
                };
                return IsLength;
            }());
            exports_17("IsLength", IsLength);
        }
    }
});
System.register("logic/logics/patterns/isPattern.logic", [], function(exports_18, context_18) {
    "use strict";
    var __moduleName = context_18 && context_18.id;
    var IsPattern;
    return {
        setters:[],
        execute: function() {
            IsPattern = (function () {
                function IsPattern() {
                }
                IsPattern.check = function (control, pattern) {
                    return pattern.test(control.value);
                };
                return IsPattern;
            }());
            exports_18("IsPattern", IsPattern);
        }
    }
});
System.register("logic/logics/patterns/isEmail.logic", ["logic/logics/patterns/isPattern.logic"], function(exports_19, context_19) {
    "use strict";
    var __moduleName = context_19 && context_19.id;
    var IsPattern_logic_1;
    var IsEmail;
    return {
        setters:[
            function (IsPattern_logic_1_1) {
                IsPattern_logic_1 = IsPattern_logic_1_1;
            }],
        execute: function() {
            IsEmail = (function () {
                function IsEmail() {
                }
                IsEmail.check = function (control) {
                    var pattern = /\S+@\S+\.\S+/;
                    return IsPattern_logic_1.IsPattern.check(control, pattern);
                };
                return IsEmail;
            }());
            exports_19("IsEmail", IsEmail);
        }
    }
});
System.register("logic", ["logic/logics/hasValue.logic", "logic/logics/numeric/isNumeric.logic", "logic/logics/numeric/min.logic", "logic/logics/numeric/max.logic", "logic/logics/numeric/range.logic", "logic/logics/length/minLength.logic", "logic/logics/length/maxLength.logic", "logic/logics/length/length.logic", "logic/logics/length/isLength.logic", "logic/logics/patterns/isPattern.logic", "logic/logics/patterns/isEmail.logic"], function(exports_20, context_20) {
    "use strict";
    var __moduleName = context_20 && context_20.id;
    return {
        setters:[
            function (hasValue_logic_1_1) {
                exports_20({
                    "HasValue": hasValue_logic_1_1["HasValue"]
                });
            },
            function (isNumeric_logic_1_1) {
                exports_20({
                    "IsNumeric": isNumeric_logic_1_1["IsNumeric"]
                });
            },
            function (min_logic_1_1) {
                exports_20({
                    "Min": min_logic_1_1["Min"]
                });
            },
            function (max_logic_1_1) {
                exports_20({
                    "Max": max_logic_1_1["Max"]
                });
            },
            function (range_logic_1_1) {
                exports_20({
                    "Range": range_logic_1_1["Range"]
                });
            },
            function (minLength_logic_1_1) {
                exports_20({
                    "MinLength": minLength_logic_1_1["MinLength"]
                });
            },
            function (maxLength_logic_1_1) {
                exports_20({
                    "MaxLength": maxLength_logic_1_1["MaxLength"]
                });
            },
            function (length_logic_1_1) {
                exports_20({
                    "Length": length_logic_1_1["Length"]
                });
            },
            function (isLength_logic_1_1) {
                exports_20({
                    "IsLength": isLength_logic_1_1["IsLength"]
                });
            },
            function (isPattern_logic_1_1) {
                exports_20({
                    "IsPattern": isPattern_logic_1_1["IsPattern"]
                });
            },
            function (isEmail_logic_1_1) {
                exports_20({
                    "IsEmail": isEmail_logic_1_1["IsEmail"]
                });
            }],
        execute: function() {
        }
    }
});
