// angular2-datatable/src/BootstrapPaginator.ts
import {Component, Input, OnChanges} from "@angular/core";
import {DataTable} from "./DataTable";
import {Paginator} from "./Paginator";
import * as _ from "lodash";

@Component({
    selector: "mfBootstrapPaginator",
    styles:[`
      .menu span.item{
          cursor: pointer;
          cursor: hand;
      }
    `],
    template: `
    <mfPaginator #p [mfTable]="mfTable">
        <div class="ui right floated pagination menu" *ngIf="p.dataLength > 0">
            <span class="icon item" [class.disabled]="p.activePage <= 1" (click)="p.setPage(1)">
               <i class="left chevron icon"></i>
            </span>
            <span class="item" *ngIf="p.activePage > 4 && p.activePage + 1 > p.lastPage" (click)="p.setPage(p.activePage - 4)">
                {{p.activePage-4}}
            </span>
            <span class="item" *ngIf="p.activePage > 3 && p.activePage + 2 > p.lastPage" (click)="p.setPage(p.activePage - 3)">
                {{p.activePage-3}}
            </span>
            <span class="item" *ngIf="p.activePage > 2" (click)="p.setPage(p.activePage - 2)">
                {{p.activePage-2}}
            </span>
            <span class="item" *ngIf="p.activePage > 1" (click)="p.setPage(p.activePage - 1)">
                {{p.activePage-1}}
            </span>
            <span class="active item">
                {{p.activePage}}
            </span>
            <span class="item" *ngIf="p.activePage + 1 <= p.lastPage" (click)="p.setPage(p.activePage + 1)">
               {{p.activePage+1}}
            </span>
            <span class="item" *ngIf="p.activePage + 2 <= p.lastPage" (click)="p.setPage(p.activePage + 2)">
                {{p.activePage+2}}
            </span>
            <span class="item" *ngIf="p.activePage + 3 <= p.lastPage && p.activePage < 3" (click)="p.setPage(p.activePage + 3)">
                {{p.activePage+3}}
            </span>
            <span class="item" *ngIf="p.activePage + 4 <= p.lastPage && p.activePage < 2" (click)="p.setPage(p.activePage + 4)">
                {{p.activePage+4}}
            </span>
            <span class="icon item" [class.disabled]="p.activePage >= p.lastPage" (click)="p.setPage(p.lastPage)">
                <i class="right chevron icon"></i>
            </span>
        </div>
        <!-- 此处的值是设置显示的行数
        <nav class="pagination pull-right" *ngIf="p.dataLength > minRowsOnPage">
            <li *ngFor="let rows of rowsOnPageSet" [class.active]="p.rowsOnPage===rows" (click)="p.setRowsOnPage(rows)">
                <span style="cursor: pointer">{{rows}}</span>
            </li>
        </nav>
        -->
    </mfPaginator>
    `,
    directives: [Paginator]
})
export class BootstrapPaginator implements OnChanges {
    @Input("rowsOnPageSet") private rowsOnPageSet = [];
    @Input("mfTable") private mfTable: DataTable;

    private minRowsOnPage = 0;

    ngOnChanges(changes:any):any {
        if(changes.rowsOnPageSet) {
            this.minRowsOnPage = _.min(this.rowsOnPageSet)
        }
    }
}
