// angular2-datatable/src/DataTable.ts
import {Directive, Input, EventEmitter, SimpleChange, OnChanges, DoCheck} from "@angular/core";
import * as _ from "lodash";

export class PageData {
  data:any;
  page: number;
  size: number;
  total :number;
  searchData:any;
  iDisplayStart:number;
  iDisplayLength:number;
}

export interface SortEvent {
    sortBy: string;
    sortOrder: string
}

export interface PageEvent {
    activePage: number;
    rowsOnPage: number;
    dataLength: number;
}

export interface DataEvent {
    length: number;
}

@Directive({
    selector: 'table[mfData]',
    outputs: ['fpage'],
    exportAs: 'mfDataTable'
})
export class DataTable implements OnChanges, DoCheck {

    @Input("mfData") public inputData:PageData = new PageData();

    private sortBy = "";
    private sortOrder = "asc";

    @Input("mfRowsOnPage") public rowsOnPage = 1000;
    @Input("mfActivePage") public activePage = 1;

    private mustRecalculateData = false;

    public data: any;
    private pageData :PageData = new PageData();

    public onDataChange = new EventEmitter<DataEvent>();
    public onSortChange = new EventEmitter<SortEvent>();
    public onPageChange = new EventEmitter<PageEvent>();
    public fpage: EventEmitter<PageData> = new EventEmitter<PageData>();

    public getSort():SortEvent {
        return {sortBy: this.sortBy, sortOrder: this.sortOrder};
    }

    public setSort(sortBy:string, sortOrder:string):void {
        if (this.sortBy !== sortBy || this.sortOrder !== sortOrder) {
            this.sortBy = sortBy;
            this.sortOrder = sortOrder;
            this.mustRecalculateData = true;
            this.onSortChange.emit({sortBy: sortBy, sortOrder: sortOrder});
        }
    }

    public getPage():PageEvent {
        return {activePage: this.inputData.page, rowsOnPage: this.inputData.size, dataLength: this.inputData.total};
    }

    public setPage(activePage:number, rowsOnPage:number):void {
        if (this.inputData.size !== rowsOnPage || this.inputData.page !== activePage) {
            this.inputData.page = this.inputData.page !== activePage ? activePage : this.calculateNewActivePage(this.inputData.size, rowsOnPage);
            this.inputData.size = rowsOnPage;
            this.mustRecalculateData = true;
            this.onPageChange.emit({activePage: this.inputData.page, rowsOnPage: this.inputData.size, dataLength: this.inputData.total});
            this.fpage.emit(this.pageData);
        }
    }

    private calculateNewActivePage(previousRowsOnPage:number, currentRowsOnPage:number):number {
        let firstRowOnPage = (this.inputData.page - 1) * previousRowsOnPage + 1;
        let newActivePage = Math.ceil(firstRowOnPage / currentRowsOnPage);
        return newActivePage;
    }

    public ngOnChanges(changes:{[key:string]:SimpleChange}):any {
        if (changes["inputData"]) {
          this.inputData = this.inputData || new PageData();
            this.onPageChange.emit({
                activePage: this.inputData.page,
                rowsOnPage: this.inputData.size,
                dataLength: this.inputData.total
            });
            this.mustRecalculateData = true;
        }
    }

    public ngDoCheck():any {  //藏检查
        if (this.mustRecalculateData) {
            this.fillData();
            this.mustRecalculateData = false;
        }
    }

    private fillData():void {
        this.inputData.page = this.inputData.page;
        this.inputData.size = this.inputData.size;

        let offset = (this.inputData.page - 1) * this.inputData.size; //开始的记录
        this.pageData.data = this.inputData.data;
        this.pageData.iDisplayStart = offset;
        this.pageData.iDisplayLength = this.inputData.size;
        this.pageData.page = this.inputData.page;
        this.pageData.size = this.inputData.size;
        //data = _.orderBy(data, [this.sortBy], [this.sortOrder]); //排序
        //data = _.slice(data, offset, offset + this.inputData.size);  //显示记录
        this.data = this.pageData;

    }
}
