// angular2-semantic-ui/components/rating/rating.js
"use strict";
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var RatingComponent = (function () {
    function RatingComponent(vm) {
        this._maxRating = 1;
        this.rating = 0;
        this.type = "star";
        this.size = "";
        this.vm = vm;
        vm.valueAccessor = this;
    }
    Object.defineProperty(RatingComponent.prototype, "maxRating", {
        get: function () {
            return this._maxRating;
        },
        set: function (val) {
            this._maxRating = val;
            if (this.rating > val) {
                this.setRating(val);
            }
            this.ratings = this.getRatings(val);
        },
        enumerable: true,
        configurable: true
    });
    RatingComponent.prototype.writeValue = function (value) {
        this.rating = value;
    };
    RatingComponent.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    RatingComponent.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    RatingComponent.prototype.ngOnInit = function () {
        this.type = this.type.toLowerCase();
        this.size = this.size.toLowerCase();
        this.ratings = this.getRatings(this.maxRating);
    };
    RatingComponent.prototype.getRatings = function (size) {
        var ratings = [];
        for (var i = 0; i < size; i++) {
            ratings.push(i + 1);
        }
        return ratings;
    };
    RatingComponent.prototype.setRating = function (item) {
        this.writeValue(item);
        this.vm.viewToModelUpdate(item);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number), 
        __metadata('design:paramtypes', [Number])
    ], RatingComponent.prototype, "maxRating", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], RatingComponent.prototype, "rating", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], RatingComponent.prototype, "type", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], RatingComponent.prototype, "size", void 0);
    RatingComponent = __decorate([
        core_1.Component({
            selector: 'lsu-rating[ngModel]',
            template: "\n    <div class=\"ui {{ type }} {{ size }} rating\">\n      <i class=\"icon\" *ngFor=\"let item of ratings\" [ngClass]=\"{'active': item <= rating }\" (click)=\"setRating(item)\"></i>\n    </div>\n  "
        }), 
        __metadata('design:paramtypes', [common_1.NgModel])
    ], RatingComponent);
    return RatingComponent;
}());
exports.RatingComponent = RatingComponent;
