// angular2-semantic-ui/components/accordion/accordion_panel.js
var core_1 = require('@angular/core');
var AccordionPanelComponent = (function () {
    function AccordionPanelComponent() {
        this.title = "";
        this.active = false;
    }
    AccordionPanelComponent.prototype.onChange = function (parent) {
        this.parent = parent;
    };
    AccordionPanelComponent.prototype.setAccordion = function () {
        this.parent.setAccordion(this);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], AccordionPanelComponent.prototype, "title", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AccordionPanelComponent.prototype, "active", void 0);
    AccordionPanelComponent = __decorate([
        core_1.Component({
            selector: 'lsu-accordionPanel',
            template: "\n    <div class=\"title\" [ngClass]=\"{'active': active}\" (click)=\"setAccordion()\">\n      <i class=\"dropdown icon\"></i>\n      {{ title }}\n    </div>\n    <div class=\"content\" [ngClass]=\"{'active': active}\">\n      <ng-content></ng-content>\n    </div>\n  "
        }), 
        __metadata('design:paramtypes', [])
    ], AccordionPanelComponent);
    return AccordionPanelComponent;
})();
exports.AccordionPanelComponent = AccordionPanelComponent;
