// ng2-semantic-ui/components/dropdown/dropdown.service.js
var core_1 = require('@angular/core');
var DISABLED = 'disabled';
var OUTSIDECLICK = 'outsideClick';
exports.KEYCODE = {
    LEFT: 37,
    UP: 38,
    RIGHT: 39,
    DOWN: 40,
    ESCAPE: 27,
    ENTER: 13,
    BACKSPACE: 8
};
var DropdownService = (function () {
    function DropdownService() {
        // State Events
        this.onToggle = new core_1.EventEmitter(false);
        this.isOpenChange = new core_1.EventEmitter(false);
        // Document Event Bindings
        this.closeDropdownBind = this.closeDropdown.bind(this);
        this.keybindFilterBind = this.keybindFilter.bind(this);
        // Keyboard Navigation
        this._selectedItem = null;
        // Classes
        this.itemClass = "item";
        this.itemSelectedClass = "selected";
        this.itemDisabledClass = "disabled";
    }
    Object.defineProperty(DropdownService.prototype, "isOpen", {
        get: function () {
            return this._isOpen;
        },
        set: function (value) {
            var _this = this;
            if (this.isDisabled) {
                value = false;
            }
            this._isOpen = value;
            if (this.isOpen) {
                this.bindDocumentEvents();
                this.selectedItem = null;
            }
            else {
                this.unbindDocumentEvents();
            }
            setTimeout(function () {
                _this.onToggle.emit(_this._isOpen);
                _this.isOpenChange.emit(_this._isOpen);
            });
        },
        enumerable: true,
        configurable: true
    });
    DropdownService.prototype.toggle = function () {
        this.isOpen = !this.isOpen;
    };
    DropdownService.prototype.bindDocumentEvents = function () {
        window.document.addEventListener('click', this.closeDropdownBind, true);
        if (!this.dropdownElement.nativeElement.parentElement.hasAttribute("suiDropdownMenu")) {
            window.document.addEventListener('keydown', this.keybindFilterBind);
        }
    };
    DropdownService.prototype.unbindDocumentEvents = function () {
        window.document.removeEventListener('click', this.closeDropdownBind, true);
        window.document.removeEventListener('keydown', this.keybindFilterBind);
    };
    DropdownService.prototype.closeDropdown = function (event) {
        //Never close the dropdown if autoClose is disabled
        if (event && this.autoClose === DISABLED) {
            return;
        }
        //Don't close the dropdown when clicking the toggle
        if (event && this.dropdownElement.nativeElement.contains(event.target) &&
            !this.menuElement.nativeElement.contains(event.target)) {
            return;
        }
        //Don't close the dropdown if expanding a nested dropdown
        if (event && this.menuElement.nativeElement.contains(event.target) &&
            event.target.hasAttribute("suiDropdown")) {
            return;
        }
        //Don't close the dropdown if clicking on any input element
        if (event && this.menuElement &&
            /input|textarea/i.test(event.target.tagName) &&
            this.menuElement.nativeElement.contains(event.target)) {
            return;
        }
        //Don't close the dropdown when clicking inside if autoClose is outsideClick
        if (event && this.autoClose === OUTSIDECLICK &&
            this.menuElement &&
            this.menuElement.nativeElement.contains(event.target)) {
            return;
        }
        //Close the dropdown
        this.isOpen = false;
    };
    DropdownService.prototype.keybindFilter = function (event) {
        if (event.which === exports.KEYCODE.ESCAPE) {
            this.isOpen = false;
            return;
        }
        if (this.isOpen &&
            ([exports.KEYCODE.ENTER, exports.KEYCODE.UP, exports.KEYCODE.RIGHT, exports.KEYCODE.DOWN, exports.KEYCODE.LEFT]
                .find(function (keyCode) { return event.which == keyCode; }))) {
            event.preventDefault();
            event.stopPropagation();
            this.keyPress(event.which);
        }
    };
    Object.defineProperty(DropdownService.prototype, "selectedItem", {
        get: function () {
            return this._selectedItem;
        },
        set: function (item) {
            if (this._selectedItem) {
                this._selectedItem.classList.remove(this.itemSelectedClass);
            }
            this._selectedItem = item;
            if (item) {
                item.classList.add(this.itemSelectedClass);
            }
        },
        enumerable: true,
        configurable: true
    });
    DropdownService.prototype.keyPress = function (keyCode) {
        //noinspection FallThroughInSwitchStatementJS
        switch (keyCode) {
            case exports.KEYCODE.DOWN:
                this.selectNextItem();
                break;
            case exports.KEYCODE.UP:
                this.selectPreviousItem();
                break;
            case exports.KEYCODE.ENTER:
                if (this.selectedItem && !this.selectedItem.hasAttribute("suiDropdown")) {
                    this.selectedItem.click();
                    this.selectedItem = null;
                    break;
                }
            //Fall through on purpose! (So enter on a nested dropdown acts as right arrow)
            case exports.KEYCODE.RIGHT:
                if (this.selectedItem && this.selectedItem.hasAttribute("suiDropdown")) {
                    this.selectedItem.click();
                    this.selectedItem = this.selectedItem.querySelector("." + this.itemClass + ":not(." + this.itemDisabledClass + ")");
                }
                break;
            case exports.KEYCODE.LEFT:
                if (this.selectedItem.parentElement != this.menuElement.nativeElement) {
                    this.selectedItem.parentElement.parentElement.click();
                    this.selectedItem = this.selectedItem.parentElement.parentElement;
                }
                break;
        }
    };
    DropdownService.prototype.selectNextItem = function () {
        if (!this.selectedItem) {
            this.selectedItem = this.menuElement.nativeElement.querySelector("." + this.itemClass + ":not(." + this.itemDisabledClass + ")");
            return;
        }
        var nextItem = this.selectedItem.nextElementSibling;
        if (nextItem) {
            this.selectedItem = nextItem;
            if (this.selectedItem.classList.contains(this.itemDisabledClass)) {
                this.selectNextItem();
            }
        }
    };
    DropdownService.prototype.selectPreviousItem = function () {
        var previousItem = this.selectedItem.previousElementSibling;
        if (previousItem) {
            this.selectedItem = previousItem;
            if (this.selectedItem.classList.contains(this.itemDisabledClass)) {
                this.selectPreviousItem();
            }
        }
    };
    return DropdownService;
})();
exports.DropdownService = DropdownService;
