// ng2-validate/core/components/validation-messages/validation-messages.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var validation_message_component_1 = require('./../validation-message/validation-message.component');
var ValidationMessagesComponent = (function () {
    function ValidationMessagesComponent() {
    }
    ValidationMessagesComponent.prototype.ngOnInit = function () {
        this.messageProperty = this.messageProperty || 'text';
    };
    Object.defineProperty(ValidationMessagesComponent.prototype, "messages", {
        get: function () {
            return this.filterMessages();
        },
        enumerable: true,
        configurable: true
    });
    ValidationMessagesComponent.prototype.filterMessages = function () {
        var result = null;
        if (this.control && this.control.errors && Object.keys(this.control.errors).length > 0) {
            result = [];
            var exemptions = this.exemptKeys ? this.exemptKeys : [];
            var _loop_1 = function(key) {
                if (!exemptions.find(function (e) { return e.toLowerCase() === key.toLowerCase(); })) {
                    result.push(this_1.control.errors[key]);
                }
            };
            var this_1 = this;
            for (var key in this.control.errors) {
                _loop_1(key);
            }
            if (result.length === 0) {
                result = null;
            }
        }
        return result;
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', common_1.AbstractControl)
    ], ValidationMessagesComponent.prototype, "control", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Array)
    ], ValidationMessagesComponent.prototype, "exemptKeys", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], ValidationMessagesComponent.prototype, "messageProperty", void 0);
    ValidationMessagesComponent = __decorate([
        core_1.Component({
            selector: 'validation-messages',
            template: "\n    <div *ngIf=\"messages\" *ngFor=\"#message of messages\">\n        <validation-message *ngIf=\"message\" [message]=\"message[messageProperty]\"></validation-message>\n    </div>",
            directives: [validation_message_component_1.ValidationMessageComponent]
        }), 
        __metadata('design:paramtypes', [])
    ], ValidationMessagesComponent);
    return ValidationMessagesComponent;
}());
exports.ValidationMessagesComponent = ValidationMessagesComponent;
