// ng2-semantic-ui/components/select/select-option.component.js
var core_1 = require('@angular/core');
var select_1 = require('../select');
var SelectOption = (function () {
    function SelectOption(host, el) {
        this.host = host;
        this.el = el;
        this.itemClass = true;
    }
    SelectOption.prototype.ngAfterContentInit = function () {
        var _this = this;
        this.HTML = this.el.nativeElement.innerHTML;
        var existingIndex = this.host.renderedOptions.findIndex(function (rO) { return rO.value == _this.value; });
        if (existingIndex != -1) {
            this.host.renderedOptions.splice(existingIndex, 1);
        }
        this.host.renderedOptions.push(this);
    };
    SelectOption.prototype.click = function (event) {
        event.stopPropagation();
        this.HTML = this.el.nativeElement.innerHTML;
        this.host.selectOption(this);
        return false;
    };
    __decorate([
        core_1.HostBinding('class.item'), 
        __metadata('design:type', Object)
    ], SelectOption.prototype, "itemClass", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], SelectOption.prototype, "value", void 0);
    __decorate([
        core_1.HostListener('click', ['$event']), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [MouseEvent]), 
        __metadata('design:returntype', Boolean)
    ], SelectOption.prototype, "click", null);
    SelectOption = __decorate([
        core_1.Component({
            selector: 'sui-select-option',
            template: "<ng-content></ng-content>"
        }), 
        __metadata('design:paramtypes', [select_1.Select, core_1.ElementRef])
    ], SelectOption);
    return SelectOption;
})();
exports.SelectOption = SelectOption;
