// angular2-semantic-ui/components/tab/tab.js
"use strict";
var core_1 = require('@angular/core');
var TabComponent = (function () {
    function TabComponent() {
        this.headerText = "";
        this.active = false;
    }
    TabComponent.prototype.ngAfterContentInit = function () {
        if (!this.headerText) {
            throw new Error("Please provider header text");
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], TabComponent.prototype, "headerText", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], TabComponent.prototype, "active", void 0);
    TabComponent = __decorate([
        core_1.Component({
            selector: "lsu-tab",
            styles: [
                ".ui.bottom.attached.tab.segment.active {\n      border-top: none;\n    }"
            ],
            template: "\n    <div class=\"ui tab\" [ngClass]=\"{'active': active, 'segment':useSegment, 'bottom':type == 'tabular', 'attached':type == 'tabular' }\" style=\"margin: 0; width: 100%;\">\n      <ng-content></ng-content>      \n    </div>\n  "
        }), 
        __metadata('design:paramtypes', [])
    ], TabComponent);
    return TabComponent;
}());
exports.TabComponent = TabComponent;
