// angular2-semantic-ui/components/dimmer/dimmer.js
var core_1 = require('@angular/core');
var DimmerComponent = (function () {
    function DimmerComponent(el) {
        this.parentEle = el.nativeElement.parentElement;
        this.dimmerCls = {
            active: this.active
        };
    }
    Object.defineProperty(DimmerComponent.prototype, "active", {
        set: function (val) {
            if (this.parentEle) {
                if (val) {
                    this.parentEle.classList.add("dimmable");
                    this.parentEle.classList.add("dimmed");
                }
                else {
                    this.parentEle.classList.remove("dimmable");
                    this.parentEle.classList.remove("dimmed");
                }
                this.dimmerCls.active = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean), 
        __metadata('design:paramtypes', [Boolean])
    ], DimmerComponent.prototype, "active", null);
    DimmerComponent = __decorate([
        core_1.Component({
            selector: 'lsu-dimmer',
            template: "\n    <div class=\"ui dimmer\" [ngClass]=\"dimmerCls\">\n      <div class=\"content\">\n        <div class=\"center\">\n          <ng-content></ng-content>\n        </div>\n      </div>\n    </div>\n  "
        }), 
        __metadata('design:paramtypes', [core_1.ElementRef])
    ], DimmerComponent);
    return DimmerComponent;
})();
exports.DimmerComponent = DimmerComponent;
