// ng2-uploader/src/directives/ng-file-select.js
var core_1 = require('@angular/core');
var ng2_uploader_1 = require('../services/ng2-uploader');
var NgFileSelect = (function () {
    function NgFileSelect(el) {
        var _this = this;
        this.el = el;
        this.onUpload = new core_1.EventEmitter();
        this.uploader = new ng2_uploader_1.Ng2Uploader();
        setTimeout(function () {
            _this.uploader.setOptions(_this.options);
        });
        this.uploader._emitter.subscribe(function (data) {
            _this.onUpload.emit(data);
        });
    }
    NgFileSelect.prototype.onFiles = function () {
        var files = this.el.nativeElement.files;
        if (files.length) {
            this.uploader.addFilesToQueue(files);
        }
    };
    NgFileSelect = __decorate([
        core_1.Directive({
            selector: '[ng-file-select]',
            inputs: ['options: ng-file-select'],
            outputs: ['onUpload'],
            host: { '(change)': 'onFiles()' }
        }), 
        __metadata('design:paramtypes', [core_1.ElementRef])
    ], NgFileSelect);
    return NgFileSelect;
})();
exports.NgFileSelect = NgFileSelect;
