// angular2-semantic-ui/components/checkbox/checkbox.js
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var CheckBoxComponent = (function () {
    function CheckBoxComponent(vm) {
        this.checked = false;
        this.disabled = false;
        this.vm = vm;
        vm.valueAccessor = this;
        this.onChange = new core_1.EventEmitter();
    }
    CheckBoxComponent.prototype.writeValue = function (value) {
        this.checked = value;
    };
    CheckBoxComponent.prototype.registerOnChange = function (fn) {
        this._onChange = fn;
    };
    CheckBoxComponent.prototype.registerOnTouched = function (fn) {
        this._onTouched = fn;
    };
    CheckBoxComponent.prototype.toggleCheck = function (event) {
        if (this.disabled) {
            return;
        }
        var value = !this.checked;
        this.writeValue(value);
        this.vm.viewToModelUpdate(value);
        this.onChange.next(event);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], CheckBoxComponent.prototype, "checked", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], CheckBoxComponent.prototype, "disabled", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], CheckBoxComponent.prototype, "type", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], CheckBoxComponent.prototype, "label", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], CheckBoxComponent.prototype, "onChange", void 0);
    CheckBoxComponent = __decorate([
        core_1.Component({
            selector: "lsu-checkbox",
            template: "\n    <div class=\"ui {{type}} checkbox\" [ngClass]=\"{'checked': checked}\">\n      <input type=\"checkbox\" [attr.checked]=\"checked? 'checked':null\" [attr.disabled]=\"disabled ? 'disabled' : null\" (click)=\"toggleCheck($event)\">\n      <label>{{ label }}</label>\n    </div>\n  "
        }), 
        __metadata('design:paramtypes', [common_1.NgModel])
    ], CheckBoxComponent);
    return CheckBoxComponent;
})();
exports.CheckBoxComponent = CheckBoxComponent;
