// ng2-semantic-ui/components/accordion/accordion-panel.component.js
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var accordion_component_1 = require('./accordion.component');
var collapse_1 = require("./../collapse");
var AccordionPanel = (function () {
    function AccordionPanel(accordion) {
        this.isOpenChange = new core_1.EventEmitter(false);
        this._isOpen = false;
        this.accordion = accordion;
    }
    Object.defineProperty(AccordionPanel.prototype, "isOpen", {
        get: function () {
            return this._isOpen;
        },
        set: function (value) {
            this._isOpen = value;
            if (value) {
                this.accordion.closeOtherPanels(this);
            }
            this.isOpenChange.emit(this._isOpen);
        },
        enumerable: true,
        configurable: true
    });
    AccordionPanel.prototype.ngOnInit = function () {
        this.accordion.addPanel(this);
    };
    AccordionPanel.prototype.ngOnDestroy = function () {
        this.accordion.removePanel(this);
    };
    AccordionPanel.prototype.toggleOpen = function (event) {
        event.preventDefault();
        if (!this.isDisabled) {
            this.isOpen = !this.isOpen;
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AccordionPanel.prototype, "isDisabled", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], AccordionPanel.prototype, "isOpen", null);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], AccordionPanel.prototype, "isOpenChange", void 0);
    AccordionPanel = __decorate([
        core_1.Component({
            selector: 'sui-accordion-panel',
            directives: [collapse_1.Collapse, common_1.NgClass],
            template: "\n<div class=\"title\" [ngClass]=\"{ active: isOpen }\" (click)=\"toggleOpen($event)\">\n    <ng-content select=\"[title]\"></ng-content>\n</div>\n<div [suiCollapse]=\"!isOpen\">\n    <div class=\"content\" [ngClass]=\"{ active: isOpen }\">\n        <ng-content select=\"[content]\"></ng-content>\n    </div>\n</div>\n",
            styles: [".content { padding: .5em 0 1em } :host:last-child .content { padding-bottom: 0 }"]
        }),
        __param(0, core_1.Inject(accordion_component_1.Accordion)), 
        __metadata('design:paramtypes', [accordion_component_1.Accordion])
    ], AccordionPanel);
    return AccordionPanel;
})();
exports.AccordionPanel = AccordionPanel;
