// ng2-semantic-ui/components/accordion/accordion.component.js
var core_1 = require('@angular/core');
var Accordion = (function () {
    function Accordion() {
        this.closeOthers = true;
        this.classes = true;
        this.panels = [];
    }
    Accordion.prototype.closeOtherPanels = function (openGroup) {
        if (!this.closeOthers) {
            return;
        }
        this.panels.forEach(function (group) {
            if (group !== openGroup) {
                group.isOpen = false;
            }
        });
    };
    Accordion.prototype.addPanel = function (group) {
        this.panels.push(group);
    };
    Accordion.prototype.removePanel = function (group) {
        var index = this.panels.indexOf(group);
        if (index !== -1) {
            this.panels.splice(index, 1);
        }
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], Accordion.prototype, "closeOthers", void 0);
    __decorate([
        core_1.HostBinding('class.ui'),
        core_1.HostBinding('class.accordion'), 
        __metadata('design:type', Object)
    ], Accordion.prototype, "classes", void 0);
    Accordion = __decorate([
        core_1.Component({
            selector: 'sui-accordion',
            template: "<ng-content></ng-content>",
            styles: ["\n/* Fix for general styling issues */\n:host {\n    display: block;\n}\n/* Fix for styled border issue */\n:host.styled sui-accordion-panel:first-child .title {\n    border-top: none\n}\n"]
        }), 
        __metadata('design:paramtypes', [])
    ], Accordion);
    return Accordion;
})();
exports.Accordion = Accordion;
