// angular2-semantic-ui/components/modal/modal.js
"use strict";
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var ModalComponent = (function () {
    function ModalComponent(vm) {
        this.options = {};
        this.vm = vm;
        vm.valueAccessor = this;
        this.id = "lsu_modal_" + Math.random();
    }
    ModalComponent.prototype.writeValue = function (value) {
        this._showModal = value;
        if (value) {
            document.body.classList.add("dimmed");
            var self = this;
            setTimeout(function () {
                var windowHeight = document.body.offsetHeight;
                var eleHeight = self.element.offsetHeight;
                var top = (windowHeight - eleHeight) / 2;
                self.element.style.top = top + 'px';
            });
        }
        else {
            document.body.classList.remove("dimmed");
        }
    };
    ModalComponent.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    ModalComponent.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    ModalComponent.prototype.ngAfterViewInit = function () {
        document.body.classList.add("dimmable");
        this.element = document.getElementById(this.id);
    };
    ModalComponent.prototype.clickContent = function (event) {
        event.stopPropagation();
    };
    ModalComponent.prototype.closeModal = function () {
        if (!this.options.closeable) {
            return;
        }
        var val = false;
        this.writeValue(val);
        this.vm.viewToModelUpdate(val);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], ModalComponent.prototype, "options", void 0);
    ModalComponent = __decorate([
        core_1.Component({
            selector: 'lsu-modal',
            styles: [
                " .trans-fadeout{\n        -webkit-transition:all 0.3s linear;\n        -moz-transition:all 0.3s linear;\n        -ms-transition:all 0.3s linear;\n        -o-transition:all 0.3s linear;\n        transition:all 0.3s linear;\n      }\n    "
            ],
            template: "\n    <div class=\"ui dimmer modals page trans-fadeout\" style=\"display: block !important\" \n      [style.visibility] = \"_showModal ? 'visible' : 'hidden'\"\n      [style.opacity] = \"_showModal ? '1' : '0'\"  \n      (click)=\"closeModal()\">\n      <div id=\"{{id}}\" class=\"ui {{options.size || ''}} {{options.type || ''}} modal active visibility\" (click)=\"clickContent($event)\">\n        <ng-content></ng-content>    \n      </div>\n    </div>\n  "
        }), 
        __metadata('design:paramtypes', [common_1.NgModel])
    ], ModalComponent);
    return ModalComponent;
}());
exports.ModalComponent = ModalComponent;
